### PR TITLE
allow using tuple variant names as function handles

### PR DESCRIPTION
**Pull request by oli-obk**

### CODE REVIEW

**Comment by oli-obk** [comment]
rebased

File: src/terminator/mod.rs [hunk 1]
```diff
@@ -238,20 +238,32 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
+                                    lvalue,
+                                    variants[disr as usize].offsets.iter().cloned().map(Size::bytes),
+                                    args,
+                                    disr,
+                                    discr_size,
+                                )?;
+                            },
+                            Layout::StructWrappedNullablePointer { .. } |
+                            Layout::RawNullablePointer { .. } => {
+                                assert_eq!(args.len(), 1);
```
Comment (solson) [inline src/terminator/mod.rs]: Is this assertion wrong? `StructWrappedNullablePointer` at least can take multiple arguments to construct. Not sure about `RawNullablePointer`.

cc @eddyb

File: src/terminator/mod.rs [hunk 2]
```diff
@@ -238,20 +238,32 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
-
-                                for (offset, (value, value_ty)) in offsets.into_iter().zip(args) {
-                                    let field_dest = dest.offset(offset);
-                                    self.write_value_to_ptr(value, field_dest, value_ty)?;
-                                }
+                                self.assign_fields(lvalue, offsets, args)?;
+                            },
+                            Layout::General { discr, ref variants, .. } => {
+                                // FIXME: report a proper error for invalid discriminants
+                                // right now we simply go into index out of bounds
```
Comment (solson) [inline src/terminator/mod.rs]: Is it possible for `disr` to be invalid? It comes from rustc's information about the type.

File: src/terminator/mod.rs (hunk 1 above)
Comment (eddyb) [inline src/terminator/mod.rs]: This is only correct for `RawNullablePointer`.

File: src/eval_context.rs [hunk 3]
```diff
@@ -1464,3 +1486,21 @@ pub fn monomorphize_field_ty<'a, 'tcx:'a >(tcx: TyCtxt<'a, 'tcx, 'tcx>, f: &ty::
 pub fn is_inhabited<'a, 'tcx: 'a>(tcx: TyCtxt<'a, 'tcx, 'tcx>, ty: Ty<'tcx>) -> bool {
     ty.uninhabited_from(&mut FxHashSet::default(), tcx).is_empty()
 }
+
+pub trait IntoValTyPair<'tcx> {
```
Comment (oli-obk) [inline src/eval_context.rs]: I can create a tuple with two variants

File: src/terminator/mod.rs (hunk 2 above)
Comment (oli-obk) [inline src/terminator/mod.rs]: I removed the comment.

File: src/terminator/mod.rs (hunk 1 above)
Comment (oli-obk) [inline src/terminator/mod.rs]: the entire code was wrong for StructWrappedNullablePointer

File: src/terminator/mod.rs [hunk 4]
```diff
@@ -238,20 +238,32 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
-                                // FIXME: don't allocate for single or dual field structs
-                                let dest = self.force_allocation(lvalue)?.to_ptr();
-
-                                for (offset, (value, value_ty)) in offsets.into_iter().zip(args) {
-                                    let field_dest = dest.offset(offset);
-                                    self.write_value_to_ptr(value, field_dest, value_ty)?;
-                                }
+                                self.assign_fields(lvalue, offsets, args)?;
+                            },
+                            Layout::General { discr, ref variants, .. } => {
```
Comment (oli-obk) [inline src/terminator/mod.rs]: renamed for clarity

File: src/eval_context.rs [hunk 5]
```diff
@@ -660,22 +696,25 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
 
         // Handle the field index for the outer non-null variant.
-        let inner_ty = match ty.sty {
+        let (inner_offset, inner_ty) = match ty.sty {
             ty::TyAdt(adt_def, substs) => {
                 let variant = &adt_def.variants[nndiscr as usize];
                 let index = discrfield[1];
                 let field = &variant.fields[index as usize];
-                field.ty(self.tcx, substs)
+                (self.get_field_offset(ty, index as usize)?, field.ty(self.tcx, substs))
```
Comment (oli-obk) [inline src/eval_context.rs]: We used to ignore the outer most type's field offset

File: tests/run-pass/tuple_like_enum_variant_constructor_struct_pointer_opt.rs [hunk 6]
```diff
@@ -4,8 +4,21 @@ struct A<'a> {
...
+#[derive(Copy, Clone, PartialEq, Debug)]
+struct B<'a>(i32, &'a i32);
+
 fn main() {
     let x = 5;
     let a = A { x: 99, y: &x };
     assert_eq!(Some(a).map(Some), Some(Some(a)));
+    let f = B;
+    assert_eq!(Some(B(42, &x)), Some(f(42, &x)));
+    // the following doesn't compile :(
```
Comment (solson) [inline tests/run-pass/tuple_like_enum_variant_constructor_struct_pointer_opt.rs]: What's the error here?

Comment (oli-obk) [inline tests/run-pass/tuple_like_enum_variant_constructor_struct_pointer_opt.rs]:
```
 --> <anon>:8:48
  |
8 |     let f: for<'a> fn(i32, &'a i32) -> B<'a> = B;
  |                                                ^ expected concrete lifetime, found bound lifetime parameter 'a
  |
  = note: expected type `fn(i32, &'a i32) -> B<'a>`
  = note:    found type `fn(i32, &i32) -> B<'_> {B<'_>::{{constructor}}}`
```

see https://is.gd/2HEXet

**Comment by oli-obk** [comment]
removed last commit

File: tests/run-pass/tuple_like_enum_variant_constructor_struct_pointer_opt.rs (hunk 6 above)
Comment (solson) [inline tests/run-pass/tuple_like_enum_variant_constructor_struct_pointer_opt.rs]: Weird!

@eddyb: Do you know anything about this?

**Pull request merged**